### ISSUE TITLE
Add fixed-point real number type 

### ISSUE THREAD

**Issue by dorisraeli** [Qiskit/openqasm#308]
Hello ✋🏻 

Many real-time systems (sometimes solely) support [fixed-point](https://en.wikipedia.org/wiki/Fixed-point_arithmetic) representation for real numbers.
Even though floating-point is usually preferable, fixed-point is sometimes needed (e.g., because of the underlying ALUs, ease of implementation, memory restrictions, or speed requirements of some of the arithmetic operators).

### What is the expected behavior?
As OQ3 should be able to represent physical circuits, I think it would be beneficial to support a fixed-point type (with specified non-negative sizes of integer and factor parts). This will allow converting a generic OQ3 circuit to a new OQ3 circuit with only fixed-point arithmetic (with some algorithm, optimization, and induced errors obviously).
I further specifically suggest using [two's complement](https://www.google.com/search?q=yea+nay+abstain&rlz=1C1GCEA_enIL928IL928&oq=yea+nay+abstain&aqs=chrome..69i57.2823j0j4&sourceid=chrome&ie=UTF-8) for sign representation.
I guess the sizes can be required in compile-time, surely it's OK for starters.

Maybe it can look something like this:
```
fixed[3,2] a = +5.5;  # a is (0)101.10
fixed[2,3] b = -1.0;  # b is (1)11.111
ufixed[2,5] c = 3.1;  # c is 11.00011
...
float d = 1.3;
fixed e = fixed[3,2](d)  # d is (0)001.01
```

IMHO this is important as the compilation transformations from floating-point to fixed-point might insert ambiguity, errors, and latency (depends on the HW and compiler obviously). Thus, it's vital to allow to express this early as possible in the process.

Open questions:

1. Should the precisions be known in compile time? I think for now it's fine to start with that.
2. What about existing functions (e.g., `cos`)? I think we should extend support (e.g., `cos`: fixed->fixed).
3. What about mixed arithmetics (floating * fixed)? I guess forbidding it for now is fine.
4. Should we convert real and floating numbers with rounding or trimming (e.g., 0.4 is 0.0110..., so is `ufixed[0,1](0.4)` the trimmed version (0).0 which is 0, or the rounded version (0).1 which is 0.5)? Note that this issue also exists in floating-point in some manner.

What do you think about this issue?

**Comment by steleman** [comment]
> This will allow converting a generic OQ3 circuit to a new OQ3 circuit with only fixed-point arithmetic (with some algorithm, optimization, and induced errors obviously).

I would like to point you to the current OQ3 Spec:

https://github.com/Qiskit/openqasm/tree/master/source/language

where you will find many answers to your questions.

What would be the advantage, and use case, of a fixed point type that introduces conversion errors by design?

**Comment by dorisraeli** [comment]
Thanks for replying so quickly! I did not quite understand your reply. 
Also, I'm familiar with the spec. The "open questions" are questions only regarding the proposed feature.

The errors are not mitigated by ignoring not adding this feature if the target systems only support fixed-point (as some do). 
As for the advantage, as I said, some control systems only support fixed-point arithmetic, so an error will be induced somewhere. I suggest allowing to express this issue in OQ3 and enable to tackle it as soon as possible, as OQ3 aims to be an MLIR. Otherwise, each compiler from OQ3 to a control system will have to handle this issue on its own.

**Comment by dorisraeli** [comment]
Obviously, if the agenda is "we only want to focus on HW systems that support floating-point arithmetic", then this feature is less important. 😅

**Comment by jakelishman** [comment]
We actually had this exact fixed-point type in the language until fairly recently (see #267), but we couldn't really come up with any actual use-cases for it, and it was mostly just making a lot of mathematical operations difficult to work with in a general sense, so we removed it to ease the burdens on implementations.  In particular, about your points:

1. the precisions would 100% have to be known at compile time, otherwise a compiler wouldn't be able to allocate register space on the hardware correctly
2. this is one of the main problems with the type; the operations such as `cos` are generally meant to be done a compile-time only (hardware _may_ allow them at run-time, but isn't required or expected to), and requiring compilers to implement full fixed-point maths libraries is an onerous ask for little benefit (that we could think of)
3. we never came up with any serious suggestions to have this work particularly well - you're going to lose significant precision one way or the other, and it would be better that programmer has to have an explicit cast, so _they_ decide which way to lose precision.  This is still a bit up-in-the-air, but may be how _most_ type interoperabilities are handled
4. as you say, this is a similar problem to general conversions to floats, and IEEE-754 specifies a few rounding modes for those that a compiler/hardware may implement.  If we were to have the `fixed` type, we'd likely just say that the rounding method must be one of a handful, but it's up to the compiler/hardware as to which one for a particular implementation.

There is a degree to which we can't hope to represent _every_ possible piece of hardware processing, so we do need to draw the line somewhere.  The classical processing capabilities of OpenQASM 3 are a lower priority than getting all the quantum representations correct, because the (very approximate) execution model we expect to have is that almost all non-trivial run-time classical operations are handled within `extern` functions, which could be defined in any programming language.

If you have a strong use-case for why we need a `fixed` type in the core language, in a way that wouldn't be suitable for an `extern` definition, we could bring it before the types working group again.

**Comment by ndebeaudrap** [comment]
Echoing @jakelishman's sentiment:

> If you have a strong use-case for why we need a ``fixed`` type in the core language, in a way that wouldn't be suitable for an ``extern`` definition, we could bring it before the types working group again.

One possible use-case would be if the real-time systems which support fixed-point operations but not floating-point operations, would be likely to arise in important roles in a quantum processing unit, in a capacity other than in representing ``angle`` values (which are one fixed-point representation in OpenQASM 3, albeit not of real numbers as such).

**Comment by dorisraeli** [comment]
I think I now understand your stance on this. 
Do we think that dynamic circuits will never include coherence time calculations? e.g.,:
```
float[32] a = estimate_phase(q1);
Rx(a/3) q2;
```
If the answer is "dynamic circuits will only require feedback, control flow and integer arithmetic", and also we do not want to be able to represent some physical experiments, then the issue is indeed unneeded.

**Comment by mbhealy** [comment]
The TSC discussed this at today's meeting. The basic feeling is that supporting fixed-point directly in the language leads to a great many complications, while at the same time all of the functionality of fixed-point numbers can be implemented using integers, which are supported, resulting in users having direct control over the rich number corner cases. At this point several of us feel that there are more productive ways to spend our time than on trying to address and justify all of the various arbitrary decisions that need to be made to fully support this. We are not opposed to adding support in future.

In your specific example above, the value `a` needs to be converted to an angle before being used as an argument to the Rx gate. The estimate_phase function might as well directly return the angle necessary. This can be accomplished using integer math and casting to an appropriate `bit` and then `angle`.

If there is anyone that is interested in adding language-level support for fixed-point numbers to OpenQASM 3 then we welcome a PR for discussion. Some things that should be addressed in such a PR:
- How exactly does the conversion from `fixed[m,f] a` to `float` of 16, 32, or 64 bits occur?
  - Can this conversion happen at run-time? What will the time and space complexity of such a conversion be?
    - Note: Negative fixed-point numbers are likely to be represented in 2's complement, while floating point mantissas are always positive.
  - What happens when fixed-point values represented with large fractional parts fall into the underflow gap of the desired float?
- How exactly does the conversion from `float` of 16, 32, or 64 bits occur to a `fixed[m,f]`?
  - Same question about run-time conversion and complexity
  - How do NaN, or Infinity convert to fixed?
  - What if the value is larger or smaller than can be represented by the fixed precision available?
- Can `fixed[m,f]` be converted to `angle[n]`? Under what circumstances and values for m, f, and n?
  - Should we automatically divide by 2*pi during the conversion? How is this accomplished?
  - Does a value very close to zero, but negative, result in the maximum `angle[n]` or the minimum `angle[n]`?
- Can `angle[n]` be converted to `fixed[m,f]`? Under what circumstances and values for m, f, and n?
  - Should we automatically multiply by 2*pi during the conversion? How is this accomplished?
- How do we cast from `fixed[m,f]` to `fixed[n,g]`? There are 9 combinations of m, f, n, and g with equality, less, or greater between corresponding pairs. 
- Should there be special notation for fixed-point literals?

**Comment by levbishop** [comment]
Closing for now, pending on any response to the above questions. 
Happy to revisit this issue, feel free to reopen if anyone wants to take up advocating for this feature.
